### PR TITLE
[Account Storage Maps] Refactor storage into V1 and V2

### PR DESCRIPTION
**Pull request by turbolent**
## Description

- Refactor the existing mixed V1/V2 code into separate files. This simplifies the logic, and more easily allows it to be feature-flagged
  - Simplify the bookkeeping: only keep a simple maps, ordered maps are unnecessary:
    We can simply sort the new indices at commit time.
- Put the V2 format behind a feature flag, and ensure that the behaviour with the flag disabled is exactly the same as it is currently. This allows us to perform a rolling/non-HCU deployment

- Expose functions in storage to schedule accounts to be migrated to V2. 
  We'll discus which accounts and when to migrate in the upcoming sync with the Execution team.

______

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/storage.go [hunk 1]
```diff
@@ -96,279 +101,142 @@ func NewStorage(ledger atree.Ledger, memoryGauge common.MemoryGauge) *Storage {
...
+			s.cacheDomainStorageMap(
+				domainStorageKey,
+				domainStorageMap,
+			)
+		}
+	}()
 
-	// Domain doesn't exist.  Return early if createIfNotExists is false.
+	if s.IsV1Account(address) ||
+		!s.Config.StorageFormatV2Enabled {
```
Comment (fxamacker) [inline runtime/storage.go:159]: Maybe we can:
- use AccountStorageV1 for accounts with domain registers
- use AccountStorageV2 for accounts with "stored" register
- use either AccountStorageV1 or AccountStorageV2 for new accounts depending on `s.Config.StorageFormatV2Enabled`

This will allow us to have consistent account format for existing accounts regardless of `s.Config.StorageFormatV2Enabled` settings.

**Review by fxamacker — COMMENTED** [review]
Nice!  Since this is DRAFT, I only left one comment about how V1 and V2 are used.

By changing how V1 and V2 are used, we can have consistent account format for existing accounts regardless of `StorageFormatV2Enabled` settings.

File: runtime/storage.go (hunk 1 above)
Comment (turbolent) [reply runtime/storage.go:159]: If I understand this correctly, are you proposing to use `AccountStorageV2` for accounts with `"stored"` register, even if the storage format V2 feature is not enabled?

If that's the case, we won't be able to roll out this feature without an HCU. If we want to roll it out without an HCU, we cannot use `AccountStorageV2` if the feature is not enabled. If the feature is not enabled, it must behave exactly like the current code, which only supports V1.

Comment (fxamacker) [reply runtime/storage.go:159]: > are you proposing to use AccountStorageV2 for accounts with "stored" register, even if the storage format V2 feature is not enabled?

Yes. 

If account is in V2 format already and feature is disabled for any reason, do we read account data in V2 format or treat it as no-data?  Or maybe we don't need to be concerned that feature is disabled after it is enabled?

> If that's the case, we won't be able to roll out this feature without an HCU. If we want to roll it out without an HCU, we cannot use AccountStorageV2 if the feature is not enabled. If the feature is not enabled, it must behave exactly like the current code, which only supports V1.

I think we can still roll out feature (disabled by default) without an HCU because the only difference is how to handle account in V2 when feature is disabled after it was enabled long enough to migrate some accounts.

When rolled out with feature disabled, all accounts are in V1 and will continue to behave as V1.  When feature is enabled, accounts (that satisfy criteria) will migrate to V2.

If feature is disabled after it is enabled, there can be a mix of V1 and V2 accounts.

I.e. If the feature becomes re-disabled, we can handle V1 as V1, handle V2 as V2, but we would stop migrating any existing V1 to V2, and use V1 for new accounts.

Comment (turbolent) [reply runtime/storage.go:159]: > If account is in V2 format already and feature is disabled for any reason, do we read account data in V2 format or treat it as no-data? Or maybe we don't need to be concerned that feature is disabled after it is enabled?

That would mean that at some point the some program was executed where the feature was enabled, and now it is turned off again. That should never occur and we should treat that as an implementation error, as we have no means to "migrate back" and I don't think we should support that to start with (we still might need to add that in case something goes wrong, but lets consider that when it's actually the case).

Comment (fxamacker) [reply runtime/storage.go:159]: > That would mean that at some point the some program was executed where the feature was enabled, and now it is turned off again. That should never occur and we should treat that as an implementation error,

OK.  In that case, maybe we can modify `IsV1Account()` to `AccountStatus()` which returns either v1, v2, or new status, and also confirm account is not v2 when feature is disabled.  This would only read one more register.

Basically, I am just looking out for any situation that might cause the account to be in an inconsistent state.

**Review by SupunS — APPROVED** [review]
I mostly reviewed for general Go code. LGTM!

**Review by fxamacker — APPROVED** [review]
Nice!  Thanks for incorporating feedback and also for adding tests for feature flag.  

I left a comment about extra register reads we can avoid for new account when `createIfNotExists` is false and V2 is enabled.

**Comment by turbolent** [comment]
@SupunS As mentioned in the description, that and the other test related to the migration are not yet adjusted, but in the follow-up PR, #3680.

@fxamacker Some of your suggestions actually already exist in #3680 – should I maybe just merge #3680 into this PR, and then we can address everything here? (I initially tried to keep the PRs small)

**Comment by fxamacker** [comment]
> @fxamacker Some of your suggestions actually already exist in https://github.com/onflow/cadence/pull/3680 – should I maybe just merge https://github.com/onflow/cadence/pull/3680 into this PR, and then we can address everything here? (I initially tried to keep the PRs small)

@turbolent Yes.  I started reviewing [PR 3680](https://github.com/onflow/cadence/pull/3680).  I can review the differences here after you merge.

File: runtime/account_storage_v1.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+type AccountStorageV1 struct {
+	ledger      atree.Ledger
+	slabStorage atree.SlabStorage
+	memoryGauge common.MemoryGauge
+
+	// newDomainStorageMapSlabIndices contains root slab index of new domain storage maps.
+	// The indices are saved using Ledger.SetValue() during Commit().
+	// Key is StorageKey{address, accountStorageKey} and value is 8-byte slab index.
+	newDomainStorageMapSlabIndices *orderedmap.OrderedMap[interpreter.StorageDomainKey, atree.SlabIndex]
```
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/account_storage_v1.go]: Improved in 3b2af1c60a43fe9e41e0b075543522677b8de3b5

File: runtime/account_storage_v2.go [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+	memoryGauge common.MemoryGauge
+
+	// cachedAccountStorageMaps is a cache of account storage maps.
+	// Key is StorageKey{address, accountStorageKey} and value is account storage map.
+	cachedAccountStorageMaps map[interpreter.StorageKey]*interpreter.AccountStorageMap
+
+	// newAccountStorageMapSlabIndices contains root slab index of new account storage maps.
+	// The indices are saved using Ledger.SetValue() during Commit().
+	// Key is StorageKey{address, accountStorageKey} and value is 8-byte slab index.
+	newAccountStorageMapSlabIndices *orderedmap.OrderedMap[interpreter.StorageKey, atree.SlabIndex]
```
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/account_storage_v2.go:42]: Good point! I had realized that for `cachedAccountStorageMaps` already in the follow-up PR https://github.com/onflow/cadence/pull/3680/files#diff-cbb560532bb42b850ff5c0ea95cc5509483a507ff423e5e2a37b67b5376a136cR36, but good point `newAccountStorageMapSlabIndices` also can be simplified 👍

[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/account_storage_v2.go:42]: Improved this in [`f309c0c` (#3680)](https://github.com/onflow/cadence/pull/3680/commits/f309c0c1ab2feef666a71c4ae85d7f1b5d12c89f)

File: runtime/config.go [hunk 4]
```diff
@@ -37,4 +37,6 @@ type Config struct {
 	CoverageReport *CoverageReport
 	// LegacyContractUpgradeEnabled enabled specifies whether to use the old parser when parsing an old contract
 	LegacyContractUpgradeEnabled bool
+	// StorageFormatV2Enabled specifies whether storage format V2 is enabled
+	StorageFormatV2Enabled bool
```
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/config.go:41]: That is yet to be decided. In #3680 there's no automatic migration at the moment

File: runtime/migrate_domain_registers_test.go [hunk 5]
```diff
@@ -74,8 +75,16 @@ func TestMigrateDomainRegisters(t *testing.T) {
...
+
 		ledger := NewTestLedger(nil, nil)
-		storage := runtime.NewStorage(ledger, nil)
+		storage := runtime.NewStorage(
+			ledger,
+			nil,
+			runtime.StorageConfig{
+				StorageFormatV2Enabled: true,
+			},
+		)
```
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/migrate_domain_registers_test.go:74]: With the merge of #3680 I think this should be addresses now. Please let me know if other cases still need to be set to `false`.

File: runtime/storage.go [hunk 6]
```diff
@@ -96,279 +101,149 @@ func NewStorage(ledger atree.Ledger, memoryGauge common.MemoryGauge) *Storage {
...
+			s.cacheDomainStorageMap(
+				domainStorageKey,
+				domainStorageMap,
+			)
+		}
+	}()
 
-		// Create new domain storage map
-		domainStorageMap := interpreter.NewDomainStorageMap(s.memoryGauge, s, atree.Address(address))
+	if !s.Config.StorageFormatV2Enabled || s.IsV1Account(address) {
```
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/storage.go:159]: Hmm, what do you think is the best way to resolve that?

**Comment by turbolent** [comment]
@fxamacker @SupunS Merged the migration-related follow-up PR #3680 into this, so everything is now in one PR. Can you please take another look? Sorry for the inconvenience

File: runtime/account_storage_v1.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+	sort.Slice(
+		slabIndices,
+		func(i, j int) bool {
+			slabIndex1 := slabIndices[i]
+			slabIndex2 := slabIndices[j]
+			domainKey1 := slabIndex1.StorageDomainKey
+			domainKey2 := slabIndex2.StorageDomainKey
+			return domainKey1.Compare(domainKey2) < 0
+		},
+	)
```
Comment (turbolent) [inline runtime/account_storage_v1.go:131]: We now keep a simple Go `map` and sort it here at commit time

File: runtime/account_storage_v2.go [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+	sort.Slice(
+		slabIndices,
+		func(i, j int) bool {
+			slabIndex1 := slabIndices[i]
+			slabIndex2 := slabIndices[j]
+			address1 := slabIndex1.Address
+			address2 := slabIndex2.Address
+			return address1.Compare(address2) < 0
+		},
+	)
```
Comment (turbolent) [inline runtime/account_storage_v2.go:195]: We now keep a simple Go `map` and sort it here at commit time

File: runtime/storage.go [hunk 9]
```diff
@@ -494,70 +366,63 @@ func (s *Storage) commit(inter *interpreter.Interpreter, commitContractUpdates b
...
-		errors.WrapPanic(func() {
-			err = s.Ledger.SetValue(
-				pair.Key.Address[:],
-				[]byte(pair.Key.Key),
-				pair.Value[:],
-			)
-		})
-		if err != nil {
-			return interpreter.WrappedExternalError(err)
+func (s *Storage) ScheduleV2MigrationForModifiedAccounts() {
```
Comment (turbolent) [inline runtime/storage.go:412]: We now expose this (and the simple function `ScheduleV2Migration` above) to allow e.g. FVM to migrate accounts. Just like in the current feature branch, the migration is performed after committing program execution changes, and before serializing all changes.

We might not end up using this function, we'll discuss it in the meeting with the Execution team.

**Review by fxamacker — APPROVED** [review]
Nice! :+1:  I just reviewed updates and new commits from PR https://github.com/onflow/cadence/pull/3680.  

To confirm my understanding of the current implementation of feature flag & migration:
- With feature flag disabled, all accounts are assumed as in V1 format and continue to be in V1 format.  Register reads & writes are consistent with deployed code.

- With feature flag enabled, 
  - V1 accounts (with write ops) continue to be stored in V1 unless `storage.ScheduleV2MigrationForModifiedAccounts()` is called before `storage.Commit()`.

  - New accounts are stored in V2 format automatically without `storage.ScheduleV2MigrationForModifiedAccounts()`.

File: runtime/account_storage_v2.go [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+		s.newAccountStorageMapSlabIndices = map[common.Address]atree.SlabIndex{}
+	}
+	s.newAccountStorageMapSlabIndices[address] = slabIndex
+}
+
+func (s *AccountStorageV2) commit() error {
+	if s.newAccountStorageMapSlabIndices == nil {
+		return nil
+	}
+
```
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/account_storage_v2.go]: Great idea! Optimized in 95933355022d1c1c4402a823135f01523eef6c12

File: runtime/account_storage_v1.go [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+	s.newDomainStorageMapSlabIndices[storageKey] = slabIndex
+
+	return domainStorageMap
+}
+
+func (s *AccountStorageV1) commit() error {
+	if s.newDomainStorageMapSlabIndices == nil {
+		return nil
+	}
+
```
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/account_storage_v1.go]: Great idea! Optimized in 95933355022d1c1c4402a823135f01523eef6c12

File: runtime/storage.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (fxamacker) [reply runtime/storage.go:159]: Here is one way to reduce register reads.

When feature is enabled,
- for V2 account, "stored" register is read.
- for V1 account and new account,
  - if `createIfNotExists` is false, "stored" register and given domain register are read.  
  - if `createIfNotExists` is true, "stored" register and given domain registers are read first.  If domain doesn't exist, more domain registers are read.

This code can probably be simplified more but it reduces register reads.
 
```Go
if !s.Config.StorageFormatV2Enabled {
    domainStorageMap = s.AccountStorageV1.GetDomainStorageMap(...)
    ...
    return domainStorageMap 
}

if s.IsV2Account(address) {
    domainStorageMap = s.AccountStorageV2.GetDomainStorageMap(...)
    ...
    return domainStorageMap
}
 
if s.HasDomainRegister(domain)
    domainStorageMap = s.AccountStorageV1.GetDomainStorageMap(...)
    ...
    return domainStorageMap
} 

if !createIfNotExists {
   return nil
}

if s.IsNewAccount(address) {
    domainStorageMap = s.AccountStorageV2.GetDomainStorageMap(...)
    ...
    return domainStorageMap
} else {
    domainStorageMap = s.AccountStorageV1.GetDomainStorageMap(...)
    ...
    return domainStorageMap
}
```

File: runtime/storage.go [hunk 12]
```diff
@@ -494,70 +366,63 @@ func (s *Storage) commit(inter *interpreter.Interpreter, commitContractUpdates b
...
 		return err
 	}
 
+	if s.Config.StorageFormatV2Enabled {
+		err = s.migrateV1AccountsToV2(inter)
+		if err != nil {
+			return err
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/storage.go]: Good idea! Gated the commit, as well in general the construction and use of it in other places, in 0dae0600bef19f3f98a81d63bb6833ce0573ca37

**Comment by turbolent** [comment]
@fxamacker exactly, great summary 👍

**Comment by turbolent** [comment]
@fxamacker exactly, great summary 👍

File: runtime/account_storage_v1.go [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+			storageDomainKey.Address[:],
+			[]byte(storageDomainKey.Domain.Identifier()),
+			slabIndex[:],
+		)
+	})
+	if err != nil {
+		return interpreter.WrappedExternalError(err)
+	}
+	return nil
+}
```
Comment (fxamacker) [inline runtime/account_storage_v1.go:190]: Maybe refactor both `AccountStorageV1.writeStorageDomainSlabIndex()` and `AccountStorageV2.writeAccountStorageSlabIndex()` by having a function to write `atree.SlabIndex` to register.  

Since the new function would be counterpart of existing `getSlabIndexFromRegisterValue()`, maybe also move these register value reading and writing functions together to another file.

**Review by fxamacker — APPROVED** [review]
Nice!  I left a minor suggestion.

BTW, what are your thoughts about reducing register reads for new accounts when `createIfNotExists` is false:
- https://github.com/onflow/cadence/pull/3678#discussion_r1844734899

I think we can avoid reading all domain registers for the scenario that doesn't need that.

File: runtime/account_storage_v1.go (hunk 13 above)
Comment (turbolent) [reply runtime/account_storage_v1.go]: Great ideas! 

> Maybe refactor both `AccountStorageV1.writeStorageDomainSlabIndex()` and `AccountStorageV2.writeAccountStorageSlabIndex()` by having a function to write `atree.SlabIndex` to register.

Done in [4addf32](https://github.com/onflow/cadence/pull/3678/commits/4addf32051aa9cf9f39e18a3d4036e6e9af1d315)


> Since the new function would be counterpart of existing `getSlabIndexFromRegisterValue()`, maybe also move these register value reading and writing functions together to another file.



Done in [4fd610a](https://github.com/onflow/cadence/pull/3678/commits/4fd610ac529be171c69c20ca024ef7e76af09d1e)

**Comment by turbolent** [comment]
@fxamacker Sorry, I had missed the comment reply due to the follow-up commits, I'll reply there

File: runtime/storage.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/storage.go:163]: It would be great to avoid unnecessary reads, even if we have to make the code a bit more complex 👍 
A few questions:
- What is `IsV2Account`? Does it only check for the V2 `stored` register? (Currently we only have `IsV1Account`. Note that I had previously optimized `IsV1Account` to check for the V2 `stored` register first)
- What is `HasDomainRegister`? Is it maybe `hasAccountStorageMap`?
- What is `IsNewAccount`? Would this determine if the account has any register? (e.g. V2 `stored` account storage map register or any of the known V1 domain storage map registers)

Also happy to have a quick sync about this, and please also feel free to open up a follow-up PR 🙏

[the remark this answers is not in the document]
Comment (fxamacker) [reply runtime/storage.go:163]: > Also happy to have a quick sync about this, and please also feel free to open up a follow-up PR 🙏

OK, happy to do that.  I will open a follow-up PR and then we can discuss it.

[the remark this answers is not in the document]
Comment (fxamacker) [reply runtime/storage.go:163]: @turbolent I think we can merge this PR to feature branch.  

I opened PR #3683 to reduce register reads. The new PR can be reviewed and merged separately.  PTAL :pray:

[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/storage.go:163]: Thank you for opening the PR, I will try to have a look tomorrow
